features/mbedtls/targets/TARGET_NUVOTON/TARGET_M480/des/des_alt.c

```diff
@@ -293,115 +296,172 @@ int mbedtls_des3_crypt_ecb( mbedtls_des3_context *ctx,
+     *
+     * Relationship of keying option and TDES H/W mode configuration
+     * 1: All three keys are independent                    ==> TDES 3-key mode (TMODE=1, 3KEYS=1)
+     * 2: K1 and K2 are independent, and K3 = K1            ==> TDES 2-key mode (TMODE=1, 3KEYS=0)
+     * 3: All three keys are identical, i.e. K1 = K2 = K3   ==> DES mode (TMODE=0)
+     *
+     * tdes_opmode is combination of TMODE/OPMODE, but TDES_Open I/F requires TMODE/OPMODE to be separate.
+     * We need to divide tdes_opmode to TMODE and OPMODE.
+     *
+     * TDES_IN_OUT_WHL_SWAP lets TDES H/W know input/output data are arranged in below for DMA transfer:
+     * 1. BE for byte sequence in word
+     * 2. BE for word sequence in double-word
+     */
+    TDES_Open(0,                                                // Channel number (0~4) 
+            enc,                                                // 0: decode, 1: encode
+            (tdes_opmode & CRPT_TDES_CTL_TMODE_Msk) ? 1 : 0,    // 0: DES, 1: TDES 
+            (keyopt == 1) ? 1 : 0,                              // 0: TDES 2-key mode, 1: TDES 3-key mode
+            tdes_opmode & CRPT_TDES_CTL_OPMODE_Msk,             // ECB/CBC/CFB/OFB/CTR
+            TDES_IN_OUT_WHL_SWAP);                              // TDES_NO_SWAP~TDES_IN_OUT_WHL_SWAP
+
+    /* Set DES/TDES keys 
+     *
+     * TDES_SetKey requires 3x2 word array. Change 3x8 byte array to 3x2 word array.
+     */
+    unsigned i;
+    uint32_t keys3x2[3][2];
+    for (i = 0; i < 3; i ++ ) {
+        keys3x2[i][0] = nu_get32_be(key[i] + 0);
+        keys3x2[i][1] = nu_get32_be(key[i] + 4);
+    }
+    TDES_SetKey(0, keys3x2);
+
     uint32_t rmn = length;
     const unsigned char *in_pos = input;
     unsigned char *out_pos = output;
-    
-    while (rmn) {
+
+    while (rmn > 0) {
         uint32_t data_len = (rmn <= MAXSIZE_DMABUF) ? rmn : MAXSIZE_DMABUF;
-        
+
         uint32_t ivh, ivl;
         ivh = nu_get32_be(iv);
         ivl = nu_get32_be(iv + 4);
         TDES_SetInitVect(0, ivh, ivl);
-    
+
         memcpy(dmabuf_in, in_pos, data_len);
-        
+
         TDES_SetDMATransfer(0, (uint32_t) dmabuf_in, (uint32_t) dmabuf_out, data_len);
-        
-        // Start enc/dec.
-        // NOTE: Don't call driver function TDES_Start in BSP because it will override TDES_CTL[3KEYS] setting.
-        CRPT->TDES_CTL |= CRPT_TDES_CTL_START_Msk | (CRYPTO_DMA_ONE_SHOT << CRPT_TDES_CTL_DMALAST_Pos);
+
+        /* Ensure memory accesses above are completed before DMA is started
+         *
+         * Replacing __DSB() with __DMB() is also OK in this case.
+         *
+         * Refer to "multi-master systems" section with DMA in:
+         * https://static.docs.arm.com/dai0321/a/DAI0321A_programming_guide_memory_barriers_for_m_profile.pdf
+         */
+        __DSB();
+        /* Start enc/dec */
+        TDES_Start(0, CRYPTO_DMA_ONE_SHOT);
         while (CRPT->TDES_STS & CRPT_TDES_STS_BUSY_Msk);
-        
+
```

Comment:
I think there should be a memory barrier after the DMA operation, too. As far as I see it would even be allowed for the compiler to re-order the `memcpy` before the `while` loop. This appears to be different from the AES code where an interrupt signals that the H/W is done.

Reply:
@hanno-arm First, I change **DES** DMA wait mechanism to interrupt in https://github.com/ARMmbed/mbed-os/pull/4925/commits/6db3e111b1f32b4862b70efd95f620e2ad25af8d to be consistent with **PRNG**/**AES**. Then I add memory barrier to guard from reordering `while` loop and `memcpy` in https://github.com/ARMmbed/mbed-os/pull/4925/commits/2de147c79c6dfedf22249cfb029f2ab7b3317fa2.

Reply:
@ccli8 Thanks, I like that the waiting-mechanisms for DES, PRNG and AES are now uniform.